File: packages/core/src/audio/mfcc.ts
import { fftPowerSpectrum, nextPowerOfTwo } from './fft.js'

/**
 * Pure-JS MFCC extraction for the speak-to-search feature (issue #279's web
 * follow-up) — the same framing/mel-filterbank/DCT-II pipeline as
 * `tools/resynth/src/resynth/mfcc.py`, ported so it can run in a browser
 * (no numpy there). Kept numerically parallel to the Python version
 * deliberately: the reference bank is extracted with the Python tool at
 * build time, and a live microphone query is extracted with this module at
 * request time — the two need to land in a comparable feature space for a
 * DTW distance between them to mean anything.
 */

export interface MfccParams {
  frameMs: number
  hopMs: number
  nMels: number
  nMfcc: number
  silenceDb: number
}

/** Defaults mirror `MfccParams` in tools/resynth/src/resynth/mfcc.py. */
export const DEFAULT_MFCC_PARAMS: MfccParams = { frameMs: 25, hopMs: 10, nMels: 40, nMfcc: 13, silenceDb: -30 }

function hzToMel(hz: number): number {
  return 2595 * Math.log10(1 + hz / 700)
}

function melToHz(mel: number): number {
  return 700 * (10 ** (mel / 2595) - 1)
}

/** RMS per non-overlapping frame; a trailing partial frame is zero-padded. */
function frameRms(samples: Float64Array, sampleRate: number, frameMs: number): Float64Array {
  const n = Math.max(1, Math.round((sampleRate * frameMs) / 1000))
  const count = Math.ceil(samples.length / n)
  const rms = new Float64Array(count)
  for (let c = 0; c < count; c += 1) {
    let sum = 0
    for (let i = 0; i < n; i += 1) {
      const idx = c * n + i
      const v = idx < samples.length ? samples[idx]! : 0
      sum += v * v
    }
    rms[c] = Math.sqrt(sum / n)
  }
  return rms
}

/**
 * The region outside leading/trailing silence, in ms — same semantics as
 * `tools/resynth/src/resynth/audio.py`'s `active_bounds_ms`: the first and
 * last frame whose RMS is above the threshold bound the active region.
 * Returns `[0, duration]` for a clip that is silent throughout.
 */
export function activeBounds(samples: Float64Array, sampleRate: number, silenceDb: number, frameMs = 10): [number, number] {
  const rms = frameRms(samples, sampleRate, frameMs)
  const threshold = 10 ** (silenceDb / 20)
  const durationMs = (1000 * samples.length) / sampleRate

  let first = -1
  let last = -1
  for (let i = 0; i < rms.length; i += 1) {
    if (rms[i]! > threshold) {
      if (first === -1) first = i
      last = i
    }
  }
  if (first === -1) return [0, durationMs]
  return [first * frameMs, Math.min(durationMs, (last + 1) * frameMs)]
}

/** Overlapping frames, one per row. A clip shorter than one frame is zero-padded to exactly one frame. */
function frameSignal(samples: Float64Array, frameLen: number, hopLen: number): number[][] {
  const n = samples.length
  if (n < frameLen) {
    const frame = new Array<number>(frameLen).fill(0)
    for (let i = 0; i < n; i += 1) frame[i] = samples[i]!
    return [frame]
  }
  const nFrames = 1 + Math.floor((n - frameLen) / hopLen)
  const frames: number[][] = []
  for (let f = 0; f < nFrames; f += 1) {
    const start = f * hopLen
    const frame = new Array<number>(frameLen)
    for (let i = 0; i < frameLen; i += 1) frame[i] = samples[start + i]!
    frames.push(frame)
  }
  return frames
}

function hammingWindow(n: number): number[] {
  if (n <= 1) return new Array<number>(n).fill(1)
  const w = new Array<number>(n)
  for (let i = 0; i < n; i += 1) w[i] = 0.54 - 0.46 * Math.cos((2 * Math.PI * i) / (n - 1))
  return w
}

/**
 * Triangular mel filterbank, `nMels` rows of `nFft / 2 + 1` weights each.
 * HTK-style construction: filter edges are placed at mel-equal spacing over
 * `[0, Nyquist]` and converted back to Hz; each bin's weight is the
 * piecewise-linear triangle evaluated at that bin's actual frequency, not
 * rounded to an integer bin index — see the Python version's mfcc.py for why
 * that rounding is the classic degenerate-filter failure mode. No Slaney-style
 * area normalization, for the same reason as the Python side: this is a
 * closed system (DTW between clips run through the same extractor), so a
 * uniform per-filter gain doesn't change any ranking.
 */
function melFilterbank(sampleRate: number, nFft: number, nMels: number): number[][] {
  const nBins = nFft / 2 + 1
  const melMax = hzToMel(sampleRate / 2)
  const melPoints: number[] = []
  for (let i = 0; i < nMels + 2; i += 1) melPoints.push((melMax * i) / (nMels + 1))
  const hzPoints = melPoints.map(melToHz)
  const binFreqs: number[] = []
  for (let k = 0; k < nBins; k += 1) binFreqs.push((sampleRate / 2) * (k / (nBins - 1)))

  const fb: number[][] = []
  for (let m = 0; m < nMels; m += 1) {
    const left = hzPoints[m]!
    const center = hzPoints[m + 1]!
    const right = hzPoints[m + 2]!
    const row = new Array<number>(nBins)
    for (let k = 0; k < nBins; k += 1) {
      const f = binFreqs[k]!
      const rising = (f - left) / Math.max(center - left, 1e-12)
      const falling = (right - f) / Math.max(right - center, 1e-12)
      row[k] = Math.max(Math.min(rising, falling), 0)
    }
    fb.push(row)
  }
  return fb
}

/** Orthonormal DCT-II basis, `nMfcc` rows of `nMels` weights: `mfcc[k] = sum_n basis[k][n] * logMel[n]`. */
function dctBasis(nMfcc: number, nMels: number): number[][] {
  const basis: number[][] = []
  for (let k = 0; k < nMfcc; k += 1) {
    const row = new Array<number>(nMels)
    for (let n = 0; n < nMels; n += 1) row[n] = Math.cos((Math.PI / nMels) * (n + 0.5) * k) * Math.sqrt(2 / nMels)
    basis.push(row)
  }
  const first = basis[0]!
  for (let n = 0; n < nMels; n += 1) first[n] = first[n]! / Math.sqrt(2)
  return basis
}

/**
 * Frame-major MFCC sequence: `[nFrames][nMfcc]`. Never empty, even for a
 * silent or sub-frame-length clip. `samples` is mono PCM in `[-1, 1]`, as
 * decoded via `AudioContext.decodeAudioData` in the browser (average the
 * channels first if the source is stereo) or `soundfile` on the Python side.
 */
export function extractMfccFromSamples(
  samples: Float32Array | Float64Array,
  sampleRate: number,
  params: MfccParams = DEFAULT_MFCC_PARAMS,
): number[][] {
  const floatSamples = samples instanceof Float64Array ? samples : Float64Array.from(samples)
  const [startMs, endMs] = activeBounds(floatSamples, sampleRate, params.silenceDb)
  const startIdx = Math.floor((startMs * sampleRate) / 1000)
  const endIdx = Math.round((endMs * sampleRate) / 1000)
  const active = floatSamples.subarray(startIdx, endIdx)

  const frameLen = Math.max(1, Math.round((sampleRate * params.frameMs) / 1000))
  const hopLen = Math.max(1, Math.round((sampleRate * params.hopMs) / 1000))
  const nFft = nextPowerOfTwo(frameLen)

  const window = hammingWindow(frameLen)
  const fb = melFilterbank(sampleRate, nFft, params.nMels)
  const basis = dctBasis(params.nMfcc, params.nMels)

  return frameSignal(active, frameLen, hopLen).map((frame) => {
    const windowed = frame.map((v, i) => v * window[i]!)
    const power = fftPowerSpectrum(windowed, nFft)
    const melEnergy = fb.map((row) => {
      let sum = 0
      for (let k = 0; k < row.length; k += 1) sum += row[k]! * (power[k]! / frameLen)
      return sum
    })
    const logMel = melEnergy.map((e) => Math.log(Math.max(e, 1e-10)))
    return basis.map((row) => row.reduce((sum, w, n) => sum + w * logMel[n]!, 0))
  })
}

File: tools/resynth/src/resynth/mfcc.py
"""MFCC extraction for the syllable classifier (issue #279).

Deliberately separate from features.py's WORLD-based pitch/duration/level
measurements: this is spectral-envelope detail for identifying *which*
syllable was said, not tone-grading a known one. A local overlapping,
windowed framing helper lives here rather than reusing audio.py's
`frame_rms` (non-overlapping, RMS-only, built for silence detection) — the
two framings serve different jobs and sharing would obscure both. Silence
*boundaries* are still detected via audio.py's `active_bounds_ms`, so a
clip's raw leading/trailing silence doesn't dilute the spectral shape DTW
compares.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from functools import lru_cache

import numpy as np

from .audio import Wave, active_bounds_ms


@dataclass(frozen=True)
class MfccParams:
    frame_ms: float = 25.0
    hop_ms: float = 10.0
    n_mels: int = 40
    n_mfcc: int = 13
    # Same semantics/default as features.AnalysisParams.silence_db — trims
    # leading/trailing silence before framing so DTW compares syllables, not
    # however much dead air happened to surround each take.
    silence_db: float = -30.0

    @classmethod
    def from_json(cls, raw: dict | None) -> "MfccParams":
        raw = raw or {}
        return cls(
            frame_ms=float(raw.get("frameMs", cls.frame_ms)),
            hop_ms=float(raw.get("hopMs", cls.hop_ms)),
            n_mels=int(raw.get("nMels", cls.n_mels)),
            n_mfcc=int(raw.get("nMfcc", cls.n_mfcc)),
            silence_db=float(raw.get("silenceDb", cls.silence_db)),
        )


def params_as_json(params: MfccParams) -> dict:
    return {k: v for k, v in asdict(params).items()}


def _next_pow2(n: int) -> int:
    return 1 << (max(n, 1) - 1).bit_length()


def _frame_signal(samples: np.ndarray, frame_len: int, hop_len: int) -> np.ndarray:
    """Overlapping frames, shape (n_frames, frame_len). A clip shorter than
    one frame is zero-padded to exactly one frame rather than yielding zero
    frames — extract_mfcc must never return an empty sequence."""
    n = len(samples)
    if n < frame_len:
        padded = np.zeros(frame_len)
        padded[:n] = samples
        return padded[np.newaxis, :]
    n_frames = 1 + (n - frame_len) // hop_len
    idx = np.arange(frame_len)[None, :] + hop_len * np.arange(n_frames)[:, None]
    return samples[idx]


def _hz_to_mel(hz: np.ndarray | float) -> np.ndarray | float:
    return 2595.0 * np.log10(1.0 + hz / 700.0)


def _mel_to_hz(mel: np.ndarray | float) -> np.ndarray | float:
    return 700.0 * (10.0 ** (mel / 2595.0) - 1.0)


@lru_cache(maxsize=8)
def _mel_filterbank(sample_rate: int, n_fft: int, n_mels: int) -> np.ndarray:
    """Triangular mel filterbank, shape (n_mels, n_fft // 2 + 1).

    Each filter's left/center/right edges are placed at mel-equal spacing
    over [0 Hz, Nyquist] and converted back to Hz (the standard HTK-style
    construction); each rfft bin's weight is the piecewise-linear triangle
    evaluated at that bin's *actual* frequency (continuous, not rounded to
    an integer bin index) — this avoids the classic degenerate-filter
    failure mode where rounding collapses a low-frequency filter to zero
    width. No Slaney-style area normalization: this is a closed system
    (DTW nearest-neighbor between clips run through the same extractor), so
    a per-filter gain that scales reference and query identically doesn't
    change any ranking — flagged as a deliberate simplicity choice, not an
    oversight.
    """
    n_bins = n_fft // 2 + 1
    mel_min, mel_max = 0.0, _hz_to_mel(sample_rate / 2.0)
    mel_points = np.linspace(mel_min, mel_max, n_mels + 2)
    hz_points = _mel_to_hz(mel_points)
    bin_freqs = np.linspace(0.0, sample_rate / 2.0, n_bins)

    fb = np.zeros((n_mels, n_bins))
    for m in range(n_mels):
        left, center, right = hz_points[m], hz_points[m + 1], hz_points[m + 2]
        rising = (bin_freqs - left) / max(center - left, 1e-12)
        falling = (right - bin_freqs) / max(right - center, 1e-12)
        fb[m] = np.clip(np.minimum(rising, falling), 0.0, None)
    return fb


@lru_cache(maxsize=8)
def _dct_basis(n_mfcc: int, n_mels: int) -> np.ndarray:
    """Orthonormal DCT-II basis, shape (n_mfcc, n_mels): `mfcc = basis @ log_mel`.

    Orthonormal scaling (the sqrt(2/N) / sqrt(1/N) factors) is the
    conventional MFCC definition but not load-bearing here either — a
    uniform per-coefficient rescale doesn't change nearest-neighbor
    ranking. Kept for familiarity/debuggability against reference MFCC
    implementations, not because the closed system needs it.
    """
    n = np.arange(n_mels)
    k = np.arange(n_mfcc).reshape(-1, 1)
    basis = np.cos(np.pi / n_mels * (n + 0.5) * k) * np.sqrt(2.0 / n_mels)
    basis[0] *= 1.0 / np.sqrt(2.0)
    return basis


def extract_mfcc(wave: Wave, params: MfccParams) -> list[list[float]]:
    """Frame-major MFCC sequence: `[n_frames][n_mfcc]`. Never empty, even for
    a silent or sub-frame-length clip (see `_frame_signal`)."""
    sr = wave.sample_rate
    start_ms, end_ms = active_bounds_ms(wave.samples, sr, params.silence_db)
    active = wave.samples[int(start_ms * sr / 1000.0) : int(round(end_ms * sr / 1000.0))]

    frame_len = max(1, round(sr * params.frame_ms / 1000.0))
    hop_len = max(1, round(sr * params.hop_ms / 1000.0))
    n_fft = _next_pow2(frame_len)

    frames = _frame_signal(active, frame_len, hop_len)
    windowed = frames * np.hamming(frame_len)[None, :]
    spectrum = np.fft.rfft(windowed, n=n_fft, axis=1)
    power = (np.abs(spectrum) ** 2) / frame_len

    fb = _mel_filterbank(sr, n_fft, params.n_mels)
    mel_energy = power @ fb.T
    log_mel = np.log(np.maximum(mel_energy, 1e-10))

    basis = _dct_basis(params.n_mfcc, params.n_mels)
    mfcc = log_mel @ basis.T
    return [[round(float(v), 6) for v in frame] for frame in mfcc]
